src/cargo/util/command_prelude.rs

```diff
@@ -1174,6 +1178,122 @@ fn get_target_triples_from_rustc() -> CargoResult<Vec<clap_complete::CompletionC
+
+    let package_map =
+        get_packages()
+            .unwrap_or_default()
+            .into_iter()
+            .fold(package_map, |mut map, package| {
+                map.entry(package.name().as_str())
+                    .or_insert_with(Vec::new)
+                    .push(package);
+                map
+            });
+
+    package_map
+        .into_iter()
+        .flat_map(|(name, packages)| {
+            // For unique package name
+            if packages.len() == 1 {
+                return vec![
+                    clap_complete::CompletionCandidate::new(name.to_string()).help(
+                        packages[0]
+                            .manifest()
+                            .metadata()
+                            .description
+                            .to_owned()
+                            .map(From::from),
+                    ),
+                ];
+            }
+
+            let version_map = HashMap::<Version, Vec<Package>>::new();
+            let version_map = packages.into_iter().fold(version_map, |mut map, package| {
+                map.entry(package.version().to_owned())
+                    .or_insert_with(Vec::new)
+                    .push(package);
+                map
+            });
+
+            version_map
+                .into_iter()
+                .flat_map(|(version, packages)| {
+                    // For package name with duplicates but unique version
+                    if packages.len() == 1 {
+                        return vec![clap_complete::CompletionCandidate::new(format!(
+                            "{}@{}",
+                            name, version
+                        ))
+                        .help(
+                            packages[0]
+                                .manifest()
+                                .metadata()
+                                .description
+                                .to_owned()
+                                .map(From::from),
+                        )];
+                    }
+
+                    // For package name with duplicates and duplicate version
+                    packages
+                        .into_iter()
+                        .map(|package| {
+                            clap_complete::CompletionCandidate::new(format!(
+                                "{}",
+                                package.package_id().to_spec()
+                            ))
+                            .help(
+                                package
+                                    .manifest()
+                                    .metadata()
+                                    .description
+                                    .to_owned()
+                                    .map(From::from),
+                            )
+                        })
+                        .collect::<Vec<_>>()
+                })
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>()
```

Comment:
1. are all of these `collect`s needed?  I thought `flat_map` accepted an `IntoIterator`
2. This feels like too complex for combinators for me

Reply:
Modify it a clearer form! Thanks!